### PR TITLE
Instruction count benchmark (6): Add frontend to drive microbenchmarks, and start README

### PR DESCRIPTION
**Pull request by robieta**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51123 Instruction count benchmark (7): Use  when collecting callgrind.
* #51122 Intermezzo: Add Timer support for callgrind repeats
* #51121 Intermezzo: quality of life tweaks to Timer. (Retain callgrind.out, change collect_baseline default, add __mul__ for FunctionCounts)
* **#50610 Instruction count benchmark (6): Add frontend to drive microbenchmarks, and start README**
* #50609 Instruction count benchmark (5): Add benchmark runner.
* #50608 Instruction count benchmark (4): Extend the benchmarks in both coverage and complexity.
* #50607 Instruction count benchmark (3): Enable TorchScript benchmarks
* #50606 Instruction count benchmark (2): Initial benchmark definitions.
* #50605 Instruction count benchmark (1): Types and interfaces
* #50604 add global_setup for C++ includes
* #50603 add utility to back test Timer

This PR adds:
1) A README. (Maybe it should be a separate PR. I can shuffle it if need be.)
Right now it's primarily focused on quick starts (running the standard suite, and defining ad-hoc benchmarks), with some notes on the high level system design after that.

2) A frontend to actually collect measurements.
This part is very much WIP. (But it needs a frontend in order to start experimenting.) 

`frontend/run.py` is in reasonable shape, though some of the exclusions for backtesting are somewhat suspect. 

`frontend/display.py` is very similar to `torch.utils.benchmark.Compare`, but richer. My plan is to move most of the core table code from `display` into the benchmark utils to both make `Compare` safer and more extensible, and reduce the complexity of this suite. However for now it's a good place to experiment.

`main.py` is again, just some code to enable testing. There is currently an `A/B` mode and a single env mode. (So we can start experimenting with wiring instruction counts into CI.)

# End result: A/B testing example
The following A/B tests 1.7 and HEAD on my devserver:
```
python main.py \
  --mode A/B \
  --A "source activate backtest_1_7" \
  --patch_a \
  --B "source activate ab_ref" \
  --display_time \
  --colorize
```

First sub-row is # of instructions, and the second is wall time. (Dimmed out if variance makes it suspect.) As you can see, there is some noise in the wall times but over such a large range of optimizations the improvement is much larger than the noise. (And simply omit `--display_time` to only show instructions.)

Screen Shot 2021-01-21 at 12 15 29 PM
Screen Shot 2021-01-21 at 12 15 56 PM
Screen Shot 2021-01-21 at 12 16 14 PM
Screen Shot 2021-01-21 at 12 16 23 PM

### CODE REVIEW

File: benchmarks/instruction_counts/README.md [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+A more detailed description is provided in later sections, however if you are
+just interested in testing changes then this section should be sufficient.
+
+#### Source command
+
+The benchmark suite is designed to be workflow agnostic. It takes a command
+to specify an environment, and as long as `import torch` works the benchmark
+should as well. For instance, in a git-worktree based workflow the source
+command would be something like
+`cd SOME_PYTORCH_ROOT && source actvate ENV_FOR_THIS_ROOT`, while in an
```
Comment (ezyang) [inline benchmarks/instruction_counts/README.md:13]: activate

**Review by wconstab — COMMENTED** [review]
oops, i did that thing where you make comments and then don't submit them as part of a review....
